### ISSUE TITLE
Generalization of ComboBox

### ISSUE THREAD

**Issue by jonathanolson** [phetsims/sun#363]
@samreid noted in the Area Model review that ideally ComboBox should be generalized to allow different layouts (and potentially different ways of highlighting options?) so that Areal Model doesn't need its own replacement (GenericLayoutSelectionNode):

<img width="313" alt="screen shot 2018-05-25 at 10 17 47 pm" src="https://user-images.githubusercontent.com/480878/40572452-d01830b0-6069-11e8-89d6-07732e7312ef.png">

I'll look into generalization, but I'd also like to consider fallback back to having things like AccessibleComboBox (so we can share the common code for a11y) and similar patterns.

Potentially we could include a layout function, and it could use the bounds of the result to position the background, etc.

**Comment by jonathanolson** [comment]
This would also include adding a `preferredWidth` to the API of ComboBox, since in Area Model it needs to match the width of the panels/boxes.

**Comment by jonathanolson** [comment]
Also will want a way to be able to precisely control touch areas as part of the layout?

<img width="312" alt="screen shot 2018-05-25 at 10 28 41 pm" src="https://user-images.githubusercontent.com/480878/40572490-0f6e674c-606b-11e8-8868-8f34398e0206.png">

**Comment by jonathanolson** [comment]
@ariel-phet noted this should not block Area Model's 1.1. Unassigning.

**Comment by terracoda** [comment]
@mbarlow12, is this issue relevant to you and me?

**Comment by jessegreenberg** [comment]
@terracoda indeed, I am not sure if 2D layouts for comboboxes was considered when we were working on them.

I may need to work on this issue in the near future for energy-skate-park.

**Comment by emily-phet** [comment]
@jessegreenberg @terracoda Is 2D comboboxes a subcategory of comboboxes. It will need a different way to navigate (I would guess being able to 'left' and 'right' and not just 'up' and 'down', correct?). If it's a significant effort to generalize, we may need to just move forward with a traditional combobox implementation and circle back around to this later. 
Or, perhaps an intermediate is to implement something that allows for traditional comboboxes and for only up/down navigation of 2D comboboxes (navigating from upper right to lower left) and address more complex interaction patterns later. Not sure if that makes sense, but just a thought.

**Comment by terracoda** [comment]
@emily-phet and @jessegreenberg, I had seen the 2D combobox, but did not consider it specifically in the combobox design pattern that I created. This will take some investigation, but my first guess is that we may want to consider the rows as groups of options or use a grid. 

There are promising new [design pattern examples](https://www.w3.org/TR/wai-aria-practices/#combobox) linked at ARIA Authoring Practices, one of which is a [popup grid](https://www.w3.org/TR/wai-aria-practices/examples/combobox/aria1.1pattern/grid-combo.html).

I haven't tested these yet, but it is exciting to see the examples.

**Comment by emily-phet** [comment]
@terracoda I'm not sure it's worth spending a lot of time right now coming up with interaction patterns for both regular and non-traditional pop-up variations. If you take a look and it seems like it will be a time-consuming task to tackle the 2D scenarios, let's hold off on that for now, and just move forward creating something for traditional comboboxes.
